_topics/en/soft-fork-activation.md

```diff
@@ -0,0 +1,462 @@
+    blocks didn't violate any of the soft fork's other rules.  For
+    example, in the block that triggered the 4 July 2015 chain split,
+    all the transactions followed the new soft fork rules---the
+    only reason miners suffered over $50,000 in losses was that the
+    upgrade required a block header contain a `3` and a non-upgraded
+    miner used a `2`.
+
+  - **Impractical parallelization:** with ISM, developers felt they
+    needed to wait for one fork to activate before a second fork could
+    be started.
+
+  - **Inability to fail:** ISM attempts didn't have expiration dates.
+    Once node software was released that waited for the activation
+    signal, all nodes from that point needed to continue monitoring for
+    that signal until activation occurred.  There was no way to decide
+    the soft fork was no longer needed.
+
+  - **Unpredictable activation times:** Not knowing the exact activation
+    time in advance meant it was difficult for protocol developers,
+    merchant system administrators, and mining pool operators to be
+    available in the hours shortly after activation occurred in case
+    there were any problems that needed a quick response.
+
+  BIP9 versionbits attempted to solve these problems by using the block
+  header version field as a bitfield.  Bits in the field were used for
+  signaling only---never to reject a block as invalid---and they could
+  be set in parallel.  Measurements were taken only once every 2,016
+  blocks, minimizing the window during which a small percentage of
+  hashrate would need to be lucky to project a false impression of 95%
+  signaling.  Finally, when the 95% signaling threshold was reached,
+  there would be a 2,016 block (roughly two week) "lock in" period until
+  activation occurred to allow people to prepare.  If the signaling
+  threshold wasn't reached by a timeout date, the attempt ended,
+  allowing the unused code to be removed from later node software
+  releases.
+
+  This activation method was first used with the soft fork that added
+  [BIP68][] consensus-enforced sequence numbers, [BIP112][]
+  `OP_CHECKSEQUENCEVERIFY`, and [BIP113][] nLockTime enforcement by
+  median time past.  The fork quickly made it to the locked-in phase,
+  and then proceeded automatically to activation.
+
+  #### [2016-7] BIP9, BIP148, and BIP91: the BIP141/143 segwit activation
+
+  The segwit soft fork was released with [BIP9][] activation
+  parameters.  A few miners quickly began signaling readiness but support
+  peaked far below the needed 95% threshold.  Some Bitcoin users felt
+  miners were illegitimately delaying a useful new feature and went on
+  to develop the mandatory activation that became [BIP148][].  The final
+  form of BIP148 specified rejecting any
+  blocks that didn't signal support for segwit starting on a
+  certain date.
+
+  BIP148 created the risk of a consensus failure.  If miners failed to
```

Comment:
I think BIP 148 proponents would say there was already a consensus failure and that BIP 148 didn't "create" the risk of a consensus failure. A significant part of the community wanted SegWit to be activated within the BIP 9(1 year) and miners were failing to activate it.

Perhaps instead of "BIP148 created the risk of a consensus failure" replace with "The presence of both BIP 9 and BIP 148 nodes on the network could have led to a consensus failure (or chain split)"

Reply:
You can use "consensus failure" in both a community of humans sense and a network of nodes running software that is potentially incompatible sense. 

When BIP 148 was drafted it was certainly a time where the community of humans disagreed on the best path forward in the face of miners failing to activate SegWit. And this could have led to a consensus failure on the network of nodes. But I don't think it is fair to put all the blame at the feet of the BIP 148 proponents. There wasn't consensus on the best path forward and so the risk of consensus failure was present pre the drafting of BIP 148.

Reply:
@michaelfolkson 

> You can use "consensus failure" in both a community of humans sense and a network of nodes running software that is potentially incompatible sense.

I've never heard "consensus failure" applied to a community of humans.  None of the first page of results on Google[1] refers to decision making by a community of humans: https://www.google.com/search?q="consensus+failure"  I have heard "failure to reach consensus" and "a loss of consensus" applied to humans, but I think the term "consensus failure" can safely be inferred to only apply to non-sentient computing.

[1] Google personalizes results for the searcher, so my results may be different than yours.  Let me know if that seems to be the case.

> The presence of both BIP 9 and BIP 148 nodes on the network could have led to a consensus failure (or chain split)"

I think you might misunderstand the problem.  Consensus failure was possible between BIP148 nodes and **any** nodes which didn't implement a policy that 100% complied with BIP148 mandatory signaling.  That included BIP141 nodes using BIP9---but it also included nodes that didn't implement segwit (e.g. Bitcoin Core before 0.13.1 and (IIRC) all versions of btcd at the time[1]).

[1] IIRC, btcd didn't add segwit code into its master branch until after segwit was activated.  I don't recall what libbitcoin did.

> I don't think it is fair to put all the blame at the feet of the BIP 148 proponents.

I'm not blaming BIP148 or its proponents, I'm describing one of the fundamental tradeoffs of a soft fork without very widespread enforcement.  There are times when it's more important to risk a consensus failure than it is to do nothing.

> There wasn't consensus on the best path forward and so the risk of consensus failure was present pre the drafting of BIP 148.

I'll edit this to say "[Software implementing] BIP148 created the risk of a consensus failure."

I agree that the seeds of a machine consensus failure were set by the disagreement of people, but the point when that became an actual risk where people could lose money was when someone started running a node with alternative consensus rules.

Reply:
> but the point when that became an actual risk where people could lose money was when someone started running a node with alternative consensus rules.

This argument could apply to anyone ever running BIP 148, UASF or BIP 8(LOT = true) software for a future soft fork. If there is even a single node on the network opposed to forced signaling then no one should run it as by doing so they risk (and are responsible for) consensus failure.

I guess it is technically true. But that position of not wanting to risk consensus failure effectively gives miners a permanent veto on all future soft forks (assuming there is a single node on the network opposed). In that case you have to weigh up the risk of consensus failure with the risk of giving miners a permanent veto of the activation of future soft forks.

Reply:
Let's say BIP 148 and UASF hadn't happened and the BIP 9(1 year) deployment for SegWit had expired and failed. If Core had released something with forced signaling (e.g BIP 149) as a follow up after the BIP 9(1 year) had expired would you say "Core developers have created the risk of a consensus failure"? That is an argument for Core never ever releasing something with forced signaling under **any** circumstances.

Reply:
@michaelfolkson 

I don't think those situations are analogous.  BIP148 was designed to "cause the existing 'segwit' deployment to activate without needing to release a new deployment."  That implies to me that its proponents weren't sure they had the economic weight necessary to enforce segwit on their own, which further implies that they didn't have the economic weight necessary to prevent consensus failures resulting from their forced signaling requirement.  In short: they were taking a gamble, and they knew it.

If Bitcoin Core ever releases a flag day or mandatory signaling soft fork that seems unlikely to gain overwhelming adoption among economic nodes before it activates (and is not an emergency), I would certainly also describe that as having created the risk of a consensus failure.  However, I believe it is possible for that project to wait for a change to have demonstrated overwhelming support and then release a flag day activation far enough in the future that, combined with almost all economic nodes already running Bitcoin Core, indicates almost everyone will be running that code by the time it activates.

(I think I've spent too much time on this argument already.  I will gladly read any response you send and try to incorporate your points into my edits, but I do not expect to respond further.  Thank you for your feedback.)

Reply:
The wording is hard, I suspect whatever wording you use someone will find fault with it. I won't take up any more of your time. Thanks for your responses.